### PR TITLE
Bump PHPStan to 1.10.20

### PR DESCRIPTION
**Pull request by samsonasik**
@TomasVotruba it seems new PHPStan cause error in tests:

```bash
There were 3 errors:

1) Rector\Tests\Php80\Rector\Switch_\ChangeSwitchToMatchRector\ChangeSwitchToMatchRectorTest::test with data set #11
Error: Call to undefined method Rector\Core\PhpParser\Printer\BetterStandardPrinter::pPHPStan_Node_AlwaysRememberedExpr()
```

I will check if there is possible fix for it.

### CODE REVIEW

**Comment by samsonasik** [comment]
The issue is came from new `AlwaysRememberedExpr` in PHPStan, ref https://github.com/phpstan/phpstan-src/commit/0cdda0b210a623ee299323da80771c2c84c602f9

/cc @ondrejmirtes

**Comment by ondrejmirtes** [comment]
You should extend the Printer from PHPStan instead of Standard: https://github.com/phpstan/phpstan-src/blob/1.11.x/src/Node/Printer/Printer.php

That's going to be future-proof for you.

**Comment by ondrejmirtes** [comment]
I'm gonna add `@api` above it for the next release.

**Comment by samsonasik** [comment]
@ondrejmirtes I got error: "Not part of BC promise" notice by extending that tho. 

Ok, I wil try as your suggestion.

**Comment by samsonasik** [comment]
@ondrejmirtes Ok, updated to extends `PHPStan\Node\Printer\Printer`

File: src/PhpParser/Printer/BetterStandardPrinter.php [hunk 1]
```diff
@@ -153,6 +154,14 @@ public function pFileWithoutNamespace(FileWithoutNamespace $fileWithoutNamespace
         return ltrim($content);
     }
 
+    /**
+     * required since PHPStan 1.10.20
+     */
+    protected function pPHPStan_Node_AlwaysRememberedExpr(AlwaysRememberedExpr $alwaysRememberedExpr): string
+    {
+        return $this->p($alwaysRememberedExpr->getExpr());
```
Comment (samsonasik) [inline src/PhpParser/Printer/BetterStandardPrinter.php]: @ondrejmirtes the original `sprintf()` cause addition of string:

```diff
-        $statement = match ($this->lexer->lookahead['type']) {
+        $statement = match (__phpstanRembered($this->lexer->lookahead['type'])) {
```

so I override it:

Comment (ondrejmirtes) [inline src/PhpParser/Printer/BetterStandardPrinter.php]: You don't need this method anymore.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/PhpParser/Printer/BetterStandardPrinter.php (hunk 1 above)
Comment (samsonasik) [reply src/PhpParser/Printer/BetterStandardPrinter.php]: without this, that cause addition of string like I commented above https://github.com/rectorphp/rector-src/pull/4305#discussion_r1235305525


```diff
-        $statement = match ($this->lexer->lookahead['type']) {
+        $statement = match (__phpstanRembered($this->lexer->lookahead['type'])) {
```

Comment (samsonasik) [reply src/PhpParser/Printer/BetterStandardPrinter.php]: I think the following code on `p()` method actually can be used:

```
        if ($node instanceof AlwaysRememberedExpr) {
            $content = parent::p($node->getExpr(), $parentFormatPreserved);
        } else {
            $content = parent::p($node, $parentFormatPreserved);
        }
```

Comment (samsonasik) [reply src/PhpParser/Printer/BetterStandardPrinter.php]: I updated to use AlwaysRememberedExpr expr on p() method https://github.com/rectorphp/rector-src/pull/4305/commits/f4b25b488cc3d74d01ef25ca8a26c0114341fa55

**Comment by samsonasik** [comment]
All checks have passed 🎉 @TomasVotruba @ondrejmirtes it is ready for review.

File: src/PhpParser/Printer/BetterStandardPrinter.php (hunk 1 above)
Comment (samsonasik) [reply src/PhpParser/Printer/BetterStandardPrinter.php]: I updated to use more general `Expr` and `VirtualNode` https://github.com/rectorphp/rector-src/pull/4305/commits/271726821a179bca663094a7d270301c0f9f4dbf

Comment (samsonasik) [reply src/PhpParser/Printer/BetterStandardPrinter.php]: I added method_exists check to ensure it has the `getExpr()` method for future proof https://github.com/rectorphp/rector-src/pull/4305/commits/d63fc71c225f3f6308c2b48099a775520283fa3b

File: src/PhpParser/Printer/BetterStandardPrinter.php [hunk 2]
```diff
@@ -155,7 +156,11 @@ public function pFileWithoutNamespace(FileWithoutNamespace $fileWithoutNamespace
 
     protected function p(Node $node, $parentFormatPreserved = false): string
     {
-        $content = parent::p($node, $parentFormatPreserved);
+        if ($node instanceof Expr && $node instanceof VirtualNode && method_exists($node, 'getExpr')) {
```
Comment (TomasVotruba) [inline src/PhpParser/Printer/BetterStandardPrinter.php:160]: Some comment here to explain it would be helpful. 

```suggestion
        if ($node instanceof VirtualNode && method_exists($node, 'getExpr')) {
```

File: src/PhpParser/Printer/BetterStandardPrinter.php [hunk 3]
```diff
@@ -38,7 +39,7 @@
  *
  * @property array<string, array{string, bool, string, null}> $insertionMap
  */
-final class BetterStandardPrinter extends Standard
+final class BetterStandardPrinter extends Printer
```
Comment (TomasVotruba) [inline src/PhpParser/Printer/BetterStandardPrinter.php:42]: Why this change? We should relly on basic printer, to avoid unnecesary coupling.

File: src/PhpParser/Printer/BetterStandardPrinter.php (hunk 2 above)
Comment (samsonasik) [reply src/PhpParser/Printer/BetterStandardPrinter.php:160]: `Expr` instance check instance is needed, `VirtualNode` can be `UnreachableStatementNode` which got Stmt

File: src/PhpParser/Printer/BetterStandardPrinter.php (hunk 3 above)
Comment (samsonasik) [reply src/PhpParser/Printer/BetterStandardPrinter.php:42]: That's per @ondrejmirtes suggestion above https://github.com/rectorphp/rector-src/pull/4305#issuecomment-1598825994

File: src/PhpParser/Printer/BetterStandardPrinter.php (hunk 2 above)
Comment (samsonasik) [reply src/PhpParser/Printer/BetterStandardPrinter.php:160]: I added comment why Expr and VirtualNode check is needed https://github.com/rectorphp/rector-src/pull/4305/commits/63e1744ba0968c84ae5da6a6bc365bc8d1e05501

File: src/PhpParser/Printer/BetterStandardPrinter.php (hunk 3 above)
Comment (TomasVotruba) [reply src/PhpParser/Printer/BetterStandardPrinter.php:42]: I see. That might fix it now, but will lock us to more coupling later. I'll see if there is a way to avoid traversing virtual nodes.

@ondrejmirtes I wish the original nodes would be untouched as before. This forces us to check what we consume from PHPStan.

Comment (samsonasik) [reply src/PhpParser/Printer/BetterStandardPrinter.php:42]: without extends PHPStan printer, we got error:

```
undefined method pPHPStan_Node_AlwaysRememberedExpr
```

per PR description.

**Comment by TomasVotruba** [comment]
Resolved in https://github.com/rectorphp/rector-src/pull/4306
